### ISSUE TITLE
Error on Login - re-use of email address?

### ISSUE THREAD

**Issue by TMM21042** [jaygoodswp/PHS#27]
When logging in to the demo page (phs.goodswp.com/demo) I used a new username but the previous email address.  When I started the demo I received this error.
![image](https://user-images.githubusercontent.com/26679593/29168126-934239bc-7d9b-11e7-9de0-00900b6d25ab.png)

I think we need to re-think the initial demo login so we don't lose visitors.  What I really want is a valid email address.  I think we should ask for any username (do not care if it has been used before) and then send a confirmation email to them at the address they provide.  The confirmation email should give them the link to the next demo page.  And they can return as many times as they want to that page to try the demo.  Although they have to go to their email account and open a message and click a link, does this seem to be the fastest and least frustrating option?

**Comment by jaygoodswp** [comment]
Right, the reason is because the email is used as the username on the external site and usernames must be unique for login purposes.  Yes, I can change the demo to do as you describe, in order to ensure that we are getting valid emails.  To solve the duplicate email problem, we can do one of two things on the external site: 1.) We can delete the user that previously used that email address OR 2.) We can change it to use the same username as the phs site.  Since it's just an example and not a real site, doing #1 would probably work the best as users do not have to be retained because it's just a sample.

**Comment by TMM21042** [comment]
OK.  Let's remove the password requirement from the demo signup, but still collect a username and email address.  We want valid, responsive email addresses, and a username (to use the first 4 letters for the PassHopr pw string).  When a visitor clicks "start my demo" an automated email is sent to them with a link to the page:  //phs.../demo/perform  where they can continue the demo.  This should be the only way to get to the page for first time demo users. 

We achieve four things this way.  We confirm the email address, we save the email address in our visitor table/database, we present (in the email) an acceptance of terms and conditions for the demo with language that suggests continuing accepts/acknowledges the terms, and we prevent against bots and one avenue of DDoS attacks.  The automated email (I will provide the message content) will give a short and simple explanation of privacy and security terms (needed since we are keeping their email address.) with a link to the demo/perform page.  
 
 For the possibility of a returning demo visitor (who either starts the demo again or refreshes an page after xx hours), we ignore the username content (it can be different to allow them to see a new 4 character fixed part of the PassHopr PW string) but we check if the email they use on the demo page has already been validated (its in our table/database) and the "start my demo" button sends them to "..../demo/perform" page to continue the demo.

At the external site login, I think we should 1) test/recognize the referring page was the demo/perform page, 2) expect the recent username and email address (overwriting the username from any prior demo), 3) test for the correct PassHopr PW string.  If the referring page is not demo/perform, then authentication should be the customary matching for a paid subscriber or trial subscriber.

Let me know if you have other thoughts.  Or there is a simpler approach.

**Comment by jaygoodswp** [comment]
For the first point, yes, I am working on the email now where they'll be able to get to demo/perform after they've confirmed the email address.  For the second point, I have a very simple email for now; just send me the content when it's ready.  For the third point, currently, returning users whose session has expired are prompted for their username and password.  Upon entering, they are sent to the demo/perform page.  I'll add some code to the /demo page to deal with returning users. Basically, it'll send them to demo/perform since they already confirmed their email address when they signed up.

For the external site, I'll add a task for that later as I don't have the subscriber portion done yet. I'll add code to overwrite the user in the case of a duplicate email address.

**Comment by jaygoodswp** [comment]
Regarding taking out the password, let's make the other changes first and then re-evaluate.  Not having a password will make it easy for hackers to impersonate users and gain access to their account.  If there's a concern about the number of fields for the demo, a lot of newer sites are removing the Confirm Password field on sign-up; the idea is that if the user did mis-typed the password on sign-up, they can always use the Forgot Password link to gain access later.

**Comment by jaygoodswp** [comment]
The email modification has been made.  I'm moving on to the other mods I spoke about above.

**Comment by TMM21042** [comment]
Demo Email Text:

Hi [username];
You're receiving this email from PassHopr.com because your email address was used to start our password hopping demonstration.  Before you proceed with the demo, please take a minute to review the information below and then click on the link at the bottom of this email to proceed with the demo.  Clicking the link to continue the demo is your consent to our terms below.

**Privacy Terms**
- we collect your email address to verify authenticity of visitors requesting a demonstration of password hopping
- we do not share or sell your email address.
- we may use your email address periodically to notify you of password hopping security, privacy, or simplicity features of importance.
- Please use the same email address if you return to the demo site more than once.  We will recognize your email address and direct you to the demo without the need for repeated email verification.

To accept these terms and continue with the demonstration, please click this link: _Link to PHS/Demo/Perform page_

**Comment by TMM21042** [comment]
Email verification is intended to occur at the time the "start my demo" button is clicked.  As of 5pm EDT 8/14, clicking on the "start my demo" button sends the visitor directly to the ../demo/perform page.  I think it should send them to a page that says "We have emailed a verification link to your, please look in your email inbox for a link to continue the demo."  

It should be that the link in the email address verification is the only way a first time visitor can get to the ../demo/perform page.  I tested it with a new (but false) email address and I went right to the ../demo/perform page.

**Comment by jaygoodswp** [comment]
Hey Tom!  I didn't have time last week to update the staging server.  So, the email modification is not yet live for you to review.  Also, we will need a test SMTP server to send the emails.  Sometimes, you can get away with using a dummy Google account to do this but Google has been cracking down recently. AWS has a service for SMTP but this will involve a monthly fee as there is no free tier for it.  Let me know which you'd like to try and I'll take care of it when I work on the project this week.

**Comment by TMM21042** [comment]
How many email accounts do you need for testing?  I have several web
domains with SMTP services and I could use one or two to set up email
accounts.

Tom

**Comment by jaygoodswp** [comment]
I would just need one account.  Below is the information needed.

User
Password
Server URL (or IP address)
Port

**Comment by jaygoodswp** [comment]
This is now working on phs.goodswp.com – it sends the email, the user clicks the link and is brought to /demo/perform.

**Comment by jaygoodswp** [comment]
OK, Tom.  So, the only two things that I did not do in here are: Remove the password field (not recommended for security reasons) and Going directly to demo/perform when email address is re-used (instead, it treats the user as new and sends the email).  The reason for the latter is because oftentimes, an email address is mistyped, and the wrong person gets the message.  You don't want to assume that just because an email address was used previously, that it belongs to the user who is registering.  As for the password, I did remove the Confirm Password field – now there's only three fields to fill in.
